File: Python/compand.py
import math
import struct
import sys

COMPRESSION_NAME = "Sound sample companding"
USAGE = "infile outfile [n]\n\n n optionally sets the bits per sample\n\n"

def get_file_length(file):
    """Get the length of the file"""
    current_pos = file.tell()
    file.seek(0, 2)  # SEEK_END
    length = file.tell()
    file.seek(current_pos, 0)  # SEEK_SET
    return length - current_pos

def compress_file(self, input_stream: FileIO, output: 'Compressor.BitFile', argc: int, argv: list[str]):
    compress = [0] * 256
           
    print(f"Compressing using {bits} bits per sample...")
    steps = 1 << (bits - 1)
    
    bit_output = BitFile(output_file, 'w')
    bit_output.output_bits(bits, 8)
    bit_output.output_bits(get_file_length(input_file), 32)
    
    for i in range(steps, 0, -1):
        value = int(128.0 * (math.pow(2.0, i / steps) - 1.0) + 0.5)
        for j in range(value, 0, -1):
            compress[j + 127] = i + steps - 1
            compress[128 - j] = steps - i
    
    while True:
        byte_data = input_file.read(1)
        if not byte_data:
            break
        c = byte_data[0]
        bit_output.output_bits(compress[c], bits)
    
    bit_output.flush()

def expand_file(self, input_stream: 'Compressor.BitFile', output_stream: FileIO, argc: int, argv: list[str]):
    expand = [0] * 256

    bit_input = BitFile(input_file, 'r')
    bits = bit_input.input_bits(8)
    print(f"Expanding using {bits} bits per sample...")
    
    steps = 1 << (bits - 1)
    last_value = 0
    
    for i in range(1, steps + 1):
        value = int(128.0 * (math.pow(2.0, i / steps) - 1.0) + 0.5)
        expand[steps + i - 1] = 128 + (value + last_value) // 2
        expand[steps - i] = 127 - (value + last_value) // 2
        last_value = value
    
    count = bit_input.input_bits(32)
    for _ in range(count):
        c = bit_input.input_bits(bits)
        output_file.write(bytes([expand[c]]))
    
    while argc > 0:
        print(f"Unused argument: {argv[0]}")
        argc -= 1
        argv = argv[1:]

File: C/compand.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include "bitio.h"
#include "errhand.h"
#include "main.h"

char *CompressionName = "Sound sample companding";
char *Usage =
"infile outfile [n]\n\n n optionally sets the bits per sample\n\n";

long get_file_length( FILE *file );

#ifndef SEEK_END
#define SEEK_END 2
#endif

#ifndef SEEK_SET
#define SEEK_SET 0
#endif

void CompressFile ( FILE *input,  BIT_FILE *output,  int argc,  char *argv[] )
{
    int compress[ 256 ];
    int steps;
    int bits;
    int value;
    int i;
    int j;
    int c;

    if ( argc-- > 0 )
        bits = atoi( *argv );
    else
        bits = 4;
    printf( "Compressing using %d bits per sample...\n", bits );
    steps = ( 1 << ( bits - 1 ) );
    OutputBits( output, (unsigned long) bits, 8 );
    OutputBits( output, (unsigned long) get_file_length( input ), 32 );
    for ( i = steps ; i > 0; i-- ) {
        value = (int)
           ( 128.0 * ( pow( 2.0, (double) i  /  steps ) - 1.0 ) + 0.5 );
        for ( j = value ; j > 0 ; j-- ) {
            compress[ j + 127 ] = i + steps - 1;
            compress[ 128 - j ] = steps - i;
        }
    }
    while ( ( c = getc( input ) ) != EOF )
        OutputBits( output, (unsigned long) compress[ c ], bits );
}


void ExpandFile ( BIT_FILE *input,  FILE *output,  int argc,  char *argv[] )
{
    int steps;
    int bits;
    int value;
    int last_value;
    int i;
    int c;
    long count;
    int expand[ 256 ];

    bits = (int) InputBits( input, 8 );
    printf( "Expanding using %d bits per sample...\n", bits );
    steps = ( 1 << ( bits - 1 ) );
    last_value = 0;
    for ( i = 1; i <= steps; i++ ) {
        value = (int)
            ( 128.0 * ( pow( 2.0, (double) i  /  steps ) - 1.0 ) + 0.5 );
        expand[ steps + i - 1 ] = 128 + ( value + last_value ) / 2;
        expand[ steps - i ] = 127 - ( value + last_value ) / 2;
        last_value = value;
    }
    for ( count = InputBits( input, 32 ) ; count > 0 ; count-- ) {
        c = (int) InputBits( input, bits );
        putc( expand[ c ], output );
    }
    while ( argc-- > 0 )
        printf( "Unused argument: %s\n", *argv++ );
}


long get_file_length ( FILE *file )
{
    long marker;
    long eof_ftell;

    marker = ftell( file );
    fseek( file, 0L, SEEK_END );
    eof_ftell = ftell( file );
    fseek( file, marker, SEEK_SET );
    return( eof_ftell - marker );
}

